src/sched/nop_asm_impl.h

```diff
@@ -0,0 +1,30 @@
+#pragma once
+
+#define _ASM_NOP asm volatile("nop")
+
+// Cannot use `do { } while (0)` here because there would be huge differences between -O0 and -O1.
+#define _2X(statement) \
+    {                  \
+        statement;     \
+        statement;     \
+    }
+
+#define _2X_NOP _2X(_ASM_NOP)
+#define _4X_NOP _2X(_2X_NOP)
+#define _8X_NOP _2X(_4X_NOP)
+#define _16X_NOP _2X(_8X_NOP)
+#define _32X_NOP _2X(_16X_NOP)
+
+// Cannot use `do { } while (0)` here because there would be huge differences between -O0 and -O1.
+// Each iter takes ~3ns on Intel x86 boosting to 2.9-3.0GHz (measured).
+// Tested on Skylake Xeon and CoffeeLake Mobile.
+// Tested on GCC 10/11 and LLVM 12.
+//
+// ~= 1us@3GHz between atomic variable checks.
+#define _CR_SPIN_FOR_ABOUT_1US_                                   \
+    {                                                             \
+        constexpr std::size_t kSelfSpinningBetweenChecks = 300;   \
```

Comment:
Move it out and make it static, we should not keep redefining this everywhere.

Reply:
Does not make sense. This is a constexpr and will be treated as an "alias" to a number. If we make it static, spaces will be allocated for it, which is unnecessary.

Reply:
You're already allocating space here, and it's stack space.

Reply:
Also I put 300 here just because it's 300 on my machine, doesn't mean it has to be constexpr. Ideally this should be benchmarked at runtime.

Reply:
> You're already allocating space here, and it's stack space.

We are not. It is not on the stack, because it is compile time constant.

Reply:
> Also I put 300 here just because it's 300 on my machine, doesn't mean it has to be constexpr. Ideally this should be benchmarked at runtime, e.g. run it 1000 times instead of 300, time it, and calibrate.

It should still be in the magnitude of microseconds. We can change the name if you think it is misleading. The point is just reducing contention, 1us or 500ns or 2us does not make much differences.

Reply:
> because it is compile time constant.

Visibility in compile time has nothing to do with storage.
I don't know if you can take the ref/addr of it, but as long as it doesn't fit into the instruction, it'll have to be stored somewhere.

Reply:
> We can change the name if you think it is misleading.

Tooling-wise I would still prefer to have an accurate 1us.
Didn't do online benchmarking before just because it's a little bit complicated as it can be mislead by heavy system load or evil scheduler.

Reply:
Will postpone it later if we really need more accurate timing.